### PR TITLE
Waiting for other programs and garbage collection when terminating receive-kill

### LINKED ISSUE #102: Kill queues garbage collection
V dockerized verzi poběží killl receive commandy v kontejneru který bude mít vlastní network interface s vlastním hostname např. `ip-10-0-41-40.ec2.internal`.
A k tomu vytvořený:
- Alarm - `Syrup syrup_kill_ip_10_0_41_40_ec2_internal queue is full`
- SQS - `syrup_kill_ip-10-0-40-102-ec2-internal`
TODO: proč to není stejné? @ujovlado v alarmu jsou podtržítka v SQS pomlčky. 

Nebude na to tedy fungovat současný [Garbage Collector ](https://github.com/keboola/queue-garbage-collector) který kontroluje jméno instance a jméno fronty.

Napadá mě řešení ve kterém by se v reakci na sigterm v kill receive (podobně jako https://github.com/keboola/syrup-queue/issues/101) commandu fronta a alarm smázly předtím než se to ukončí. Nereagovalo by to ale na tvrdé vypnutí instance, to by se ale teoreticky vůbec dít nemělo, u spotů reagujeme na notifikaci o plánovaném vypnutí a instanci v ECS drainujeme.

### PR DESCRIPTION
**Pull request by ujovlado**
Fixes #102 
Based on: #107 

Changes:

- when `QUEUE_RECEIVE_KILL_WAIT_FOR_OTHER_PROGRAMS_TO_END=1` the receive-kill will not stop processing kill queue until its running as last supervisor program
- when `QUEUE_RECEIVE_KILL_GARBAGE_COLLECT=1` the receive-kill will delete alarm and queue at the end (after receiving sigterm)

### CODE REVIEW

**Comment by ujovlado** [comment]
Takto sa ukonci api:

Screenshot_2019-06-10_20-27-09

A takto worker:

Screenshot_2019-06-10_20-29-27


^ tu je dobre vsimnut si `Will wait for other programs to finish`

**Comment by ujovlado** [comment]
Z PT pocas deploya na testing, oba alarmy a queue zmazane 🎉 

Screenshot 2019-06-10 at 23 18 22

File: Command/ReceiveKillCommand.php [hunk 1]
```diff
@@ -59,12 +63,100 @@ protected function execute(InputInterface $input, OutputInterface $output)
...
+                $this->output->writeln("Will wait for other programs to finish");
+            } else {
+                $this->handleEnd();
+            }
+        });
+
+        // start receiving messages from SQS
         do {
+            pcntl_signal_dispatch();
+            if ($waitForOtherProgramsToEnd === true && $this->getNumberOfProcesses() === 1) {
```
Comment (Halama) [inline Command/ReceiveKillCommand.php]: Možná by to tady mohlo logovat na kolik procesů to čeká pokud to čeká? Bojím se že to bude špatně debugovatelné. Asi to ale nesmí logovat při každé kontrole protože by to mohlo být dost verbose?

Comment (ujovlado) [reply Command/ReceiveKillCommand.php]: Nejake logovanie by tam slo dat. Tam ale uvidis asi len to, kolko `queue-receive` commandov tam bezi.

**Comment by ujovlado** [comment]
Len pre info, oba PR uz su nasadene.

**Pull request merged**